### PR TITLE
option to limit access to an IAM group instead of all IAM users

### PR DESCRIPTION
**Pull request by dylansmith**

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -13,6 +13,11 @@ cd $tmpdir/aws-ec2-ssh
 cp authorized_keys_command.sh /opt/authorized_keys_command.sh
 cp import_users.sh /opt/import_users.sh
 
+# To control which users are given access, uncomment the line below changing
+# GROUPNAME to the name of the IAM group for users. If you leave it blank, all
```
Comment (michaelwittig) [inline install.sh]: blank seems to be not correct. you check for ##ALL##

File: import_users.sh [hunk 2]
```diff
@@ -1,5 +1,18 @@
 #!/bin/bash
 
+# Specify an IAM group for users who should be given access, or leave this with
+# the value '##ALL##' to give access to all IAM users.
+UsersGroup="##ALL##"
```
Comment (michaelwittig) [inline import_users.sh]: why not blank?

Comment (dylansmith) [inline import_users.sh]: For consistency?

Comment (michaelwittig) [inline import_users.sh]: I see. so it needs to be changed in the install.sh accordingly

**Comment by shinenelson** [comment]
Aha! I was just going to open a pull request with this feature. I've implemented this in a slightly different way though. My implementation can accept multiple IAM Groups.

Check out my [import-user-from-groups](https://github.com/shinenelson/aws-ec2-ssh/tree/import-user-from-groups) branch, if you're interested.

**Comment by michaelwittig** [comment]
Hi! What do you think about putting the group membership check into the authorized_keys_command.sh script? This would eliminate the issue with group member ship changes and will also apply immediately

**Comment by shinenelson** [comment]
I'd like importing users from groups so that I can fine-tune what kind of privileges that I can give them on the servers. I'd like to put one group in a `chroot` or restrict others to an `lshell`, etc. The import process and the `useradd` command is the best place to achieve that.

But yes, it is appropriate to `authorized_keys_command` too. I'll open a pull request in a while. :wink:

**Comment by michaelwittig** [comment]
I see, The two requirements are:
* allow only a subset of users to login
* configure local users depending in the group membership

The first one seems simpler by checking for the group membership in authorized_keys_command.
The second one seems more complicated because the user needs to configure group restrictions outside of IAM I guess.

**Comment by shinenelson** [comment]
I just opened pull request #23 that implemented the authorization strategy from AWS IAM groups from with the `authorized_keys_command`.

> The second one seems more complicated because the user needs to configure group restrictions outside of IAM I guess.

That part seems pretty straight-forward with my implementation in my [import-user-from-groups](https://github.com/shinenelson/aws-ec2-ssh/blob/import-user-from-groups/import_users.sh#L5) branch. I'm using the same on my production environment as well. I have a lot of users on my account that don't need to have access to any of my EC2 Instances. What's the point of creating users for them on the instances if they're not going to be used? (that's partly a security concern as well).

**Comment by mvanbaak** [comment]
Putting it in the `authorized_keys_command` makes logging in to the instance even slower than it already is. While it has it's advantage (like michaelwittig said) it has, in my opinion, a very big downside. All the API calls done in the shellscript while logging in makes it too slow. I tried and got a lot of complaints from developers that they thought the instances were acting up.

I like the way it is done in the `import-user-from-groups` branch better.

**Comment by michaelwittig** [comment]
@shinenelson Is their a open PR with your import-user-from-groups branch? Maybe I'm missing something?

**Comment by shinenelson** [comment]
> @shinenelson Is their a open PR with your import-user-from-groups branch? Maybe I'm missing something?

No, I just mentioned it in https://github.com/widdix/aws-ec2-ssh/pull/19#issuecomment-279363236 since this pull request was already open and you had already commented an alternate strategy. However, I did open pull request #23 following your suggestion.

**Comment by michaelwittig** [comment]
#24 was merged as discussed in #28

**Pull request closed without merging**